src/fetchBaseQuery.ts

```diff
@@ -1,28 +1,73 @@
 import { QueryApi } from './buildThunks';
+import { isPlainObject, joinUrls } from './utils';
 
 interface FetchArgs extends RequestInit {
   url: string;
+  params?: URLSearchParams | Record<string, string>;
+  body?: any;
+  responseHandler?: 'json' | 'text' | ((response: Response) => Promise<any>);
+  validateStatus?: (response: Response, body: any) => boolean;
 }
 
-export function fetchBaseQuery({ baseUrl }: { baseUrl: string } = { baseUrl: '' }) {
+const defaultValidateStatus = (response: Response) => response.status >= 200 && response.status <= 299;
+
+const isJsonContentType = (headers: Headers) => headers.get('content-type')?.trim()?.startsWith('application/json');
+
+// note: and fetchBaseQuery should probably have a prepareHeaders method option --- baseQuery has access to thunkApi aka (dispatch, getState)
+export function fetchBaseQuery({ baseUrl }: { baseUrl?: string } = {}) {
   return async (arg: string | FetchArgs, { signal, rejectWithValue }: QueryApi) => {
-    const { url, method = 'GET', ...rest } = typeof arg == 'string' ? { url: arg } : arg;
-    const result = await fetch(`${baseUrl}/${url}`, {
+    let {
+      url,
+      method = 'GET' as const,
+      headers = undefined,
+      body = undefined,
+      params = undefined,
+      responseHandler = 'json' as const,
+      validateStatus = defaultValidateStatus,
+      ...rest
+    } = typeof arg == 'string' ? { url: arg } : arg;
+    let config: RequestInit = {
       method,
-      headers: {
-        'Content-Type': 'application/json',
-      },
       signal,
+      body,
       ...rest,
-    });
+    };
+
+    config.headers = new Headers(headers);
+
+    if (!config.headers.has('content-type')) {
+      config.headers.set('content-type', 'application/json');
+    }
+
+    if (body && isPlainObject(body) && isJsonContentType(config.headers)) {
+      config.body = JSON.stringify(body);
+    }
+
+    url = joinUrls(baseUrl, url);
+
+    if (params) {
+      const searchParams = new URLSearchParams(params);
+      url += `?${searchParams.toString()}`;
```

Comment:
This will lead to a double `?` if the `url` already has search params.
Maybe something like this?
```suggestion
      let parsedUrl = new URL(url);
      for ([k,v] of Object.entries(params)) { url.searchParams.set(k,v) }
      url = parsedUrl.toString()
```

Reply:
I've updated this to behave like `axios`/`redaxios` do. The problem with `new URL` is that if you give it a relative url, it requires a baseUrl, which is a major headache especially if someone tried to use this with SSR? We experimented with it yesterday and it was a ton of bloat to get something that didn't even cover every case :(